design-proposals/002-filter-data.md

```diff
@@ -0,0 +1,142 @@
+
+For the typed definition of a filter, a **filter step** contains one or more **filter rules** which are combined with _and_ or _or_
+
+A _filter step_ is a regular `Step` with a type "filter". 
+_Filter rules_ are part of the configuration of this step and stored in the properties of a step.
+The lowest building block is a a `FilterRule`. It has the following properties:
+
+* **type** : On which part of the message to apply the filter. Can be either "header", "body" or "properties".
+* **key** : The key to use to extract a message value (e.g. header or property name). The key is empty when type is equals to body.
+* **op** : The operator to use for the filter
+* **value** : The value used by the operator to determine whether a filter applies.
+* **combinator** : How this rule is combined with the previous rule. The value can be either `&&` or `||`. The first filter rule's combinator is ignored when building up the expression.
+ 
+"type" and "key" could also be combined to a selector (which e.g. then would have the format `type.key`)
+
+The configuration of a `FilterStep` is then a list of `FilterRules`.
+
+For example, the following configuration
+
+```yaml
+- type: "body"
+  op: "contains"
+  value: "antman"
+- type: "header"    
+  key: "region"
+  op: "=~"
+  value: "asia"
+  combinator: "&&"
+- type: "body"
+  op: "regex"
+  value: "bat(wo)?man"
+  combinator: "||"
+- type: "header"
+  key: "publisher"
+  op: "=~"
+  value: "DC Comics"
+  combinator: "&&"
+```
+
+would be stored as a property "filter" on the integration step. It translates later to simple lang expression
+
+```
+${body} contains "antman" && ${in.header.region} =~ "asia" || ${body} regex "bat(wo)?man" && ${in.header.publisher} =~ "DC Comics"
+```
+
+The simple expression language [does not support parentheses](http://camel.apache.org/simple.html) nor precedence of operators so the expression is always evaluated from left to right.
+
+#### Text based filter (freeform)
+
+It should be possible to add a simple expression directly as text. Ideally there should be some sort of intellisense, but in the first step its a plain text input. However the expression needs to be validated before its stored in the DB. In the examples above the simple expression would be added directly. 
+
+### Persistence
+
+Currently the configuration of a step is a plain properties map object in JSON. Each FilterStep supports the following properties:
+
+* **type** : Either "form" or "text" for the two modes described above.
+* **simple** : The simple filter expression to apply.
+* **rules** : The rules as defined above for a "form" typed integration.
+
+In the case of a "form" filter, the "simple" expression is calculated when the step is persisted.
+
+Example for a persistent integration step:
+
+```json
+[{ 
+  "id": "1",
+  "stepKind": "filter",
+  "configuredProperties": {
+    "type": "form",
+    "simple" : "${body} contains \"antman\" || ${in.header.publisher} =~ \"DC Comics\"",
```

Comment:
I think it's easier for the generation of the actual step for the runtime function.yaml to no replicate the logic to build up the simple expression (which we need when we storing the rule to validate the expression). Also, the runtime generator then doesn't care whether the expression has been created as free text or via a form and just picks the content of this field.

The price is that we add some denormalization and the possibility for some inconsistencies.